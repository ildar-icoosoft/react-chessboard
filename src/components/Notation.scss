.notation {
  font-family: "Helvetica Neue";
  z-index: 3;
  position: absolute;
  font-size: 10px;
  color: rgb(240, 217, 181);
}

.fileNotation {
  @extend .notation;
  align-self: flex-end;
}
.rankNotation {
  @extend .notation;
  align-self: flex-start;
}

.lightSquare {
  color: rgb(181, 136, 99);
}
