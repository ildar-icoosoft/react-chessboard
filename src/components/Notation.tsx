import React, { FC } from "react";
import { PieceColor } from "../enums/PieceColor";
import { getFileIndex, getRankIndex, isLightSquare } from "../utils/chess";
import css from "./Notation.scss";
import classNames from "classnames";
import { DEFAULT_SQUARE_WIDTH } from "../constants/constants";

export interface NotationProps {
  coordinates: string;
  orientation: PieceColor;
  width?: number;
}

export const Notation: FC<NotationProps> = ({
  coordinates,
  orientation,
  width = DEFAULT_SQUARE_WIDTH,
}) => {
  const rankIndex: number = getRankIndex(coordinates);
  const fileIndex: number = getFileIndex(coordinates);

  const isColumn: boolean =
    orientation === PieceColor.WHITE ? rankIndex === 0 : rankIndex === 7;
  const isRow: boolean =
    orientation === PieceColor.WHITE ? fileIndex === 0 : fileIndex === 7;

  if (isColumn && isRow) {
    return renderBottomLeft(coordinates, width);
  }

  if (isColumn) {
    return renderFile(coordinates, width);
  }

  if (isRow) {
    return renderRank(coordinates, width);
  }

  return null;
};

const renderBottomLeft = (coordinates: string, width: number) => {
  return (
    <>
      {renderRank(coordinates, width)}
      {renderFile(coordinates, width)}
    </>
  );
};

const renderFile = (coordinates: string, width: number) => {
  return (
    <div
      data-testid={`notation-file-${coordinates[0]}`}
      style={{
        fontSize: `${width / 6}px`,
        paddingLeft: `${width - width / 6}px`,
      }}
      className={classNames(css.fileNotation, {
        [css.lightSquare]: isLightSquare(coordinates),
      })}
    >
      {coordinates[0]}
    </div>
  );
};

const renderRank = (coordinates: string, width: number) => {
  return (
    <div
      data-testid={`notation-rank-${coordinates[1]}`}
      style={{
        fontSize: `${width / 6}px`,
        paddingRight: `${width - width / 6}px`,
      }}
      className={classNames(css.rankNotation, {
        [css.lightSquare]: isLightSquare(coordinates),
      })}
    >
      {coordinates[1]}
    </div>
  );
};
