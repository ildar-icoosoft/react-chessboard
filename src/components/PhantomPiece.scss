.phantomPiece {
  position: absolute;
  top: 0;
  left: 0;
  width: 12.5%;
  height: 12.5%;
  z-index: 1;
}
