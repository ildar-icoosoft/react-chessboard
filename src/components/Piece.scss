.piece {
  position: absolute;
  top: 0;
  left: 0;
  width: 12.5%;
  height: 12.5%;
  background-size: cover;
  z-index: 2;
  will-change: transform;
  pointer-events: none;
}

.bB {
  background-image: url(../images/pieces/black_bishop.svg);
}
.bK {
  background-image: url(../images/pieces/black_king.svg);
}
.bN {
  background-image: url(../images/pieces/black_knight.svg);
}
.bP {
  background-image: url(../images/pieces/black_pawn.svg);
}
.bQ {
  background-image: url(../images/pieces/black_queen.svg);
}
.bR {
  background-image: url(../images/pieces/black_rook.svg);
}

.wB {
  background-image: url(../images/pieces/white_bishop.svg);
}
.wK {
  background-image: url(../images/pieces/white_king.svg);
}
.wN {
  background-image: url(../images/pieces/white_knight.svg);
}
.wP {
  background-image: url(../images/pieces/white_pawn.svg);
}
.wQ {
  background-image: url(../images/pieces/white_queen.svg);
}
.wR {
  background-image: url(../images/pieces/white_rook.svg);
}
