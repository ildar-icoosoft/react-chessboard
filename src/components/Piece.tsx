import React, { CSSProperties, FC, useEffect, useState } from "react";
import { PieceCode } from "../enums/PieceCode";
import { Transition } from "react-transition-group";
import { SquareTransitionFrom } from "../interfaces/SquareTransitionFrom";
import { TransitionStatus } from "react-transition-group/Transition";
import { PartialRecord } from "../types/PartialRecord";
import { DEFAULT_TRANSITION_DURATION } from "../constants/constants";
import classNames from "classnames";
import css from "./Piece.scss";
import { XYCoordinates } from "../interfaces/XYCoordinates";

export interface PieceProps {
  pieceCode: PieceCode;
  width?: number;
  transitionFrom?: SquareTransitionFrom;
  transitionDuration?: number;
  xYCoordinates?: XYCoordinates;
}

const getDefaultStyle = (
  xYCoordinates: XYCoordinates | undefined
): CSSProperties => {
  if (!xYCoordinates) {
    return {};
  }
  return {
    transform: `translate(${xYCoordinates.x}px, ${xYCoordinates.y}px)`,
  };
};

const getTransitionStyles = (
  transitionFrom: SquareTransitionFrom | undefined,
  state: TransitionStatus,
  transitionDuration: number
): CSSProperties => {
  if (!transitionFrom) {
    return {};
  }

  const styles: PartialRecord<TransitionStatus, CSSProperties> = {
    exited: {
      transform: `translate(${transitionFrom.x}px, ${transitionFrom.y}px)`,
    },
    entering: {
      transform: "translate(0, 0)",
      transition: `transform ${transitionDuration}ms`,
      zIndex: 1,
    },
  };

  return styles[state] || {};
};

export const Piece: FC<PieceProps> = ({
  pieceCode,
  transitionFrom,
  transitionDuration = DEFAULT_TRANSITION_DURATION,
  xYCoordinates,
}) => {
  const [inProp, setInProp] = useState<boolean>(false);

  useEffect(() => {
    if (transitionFrom) {
      setInProp(true);
    }
  }, [transitionFrom]);

  return (
    <Transition in={inProp} timeout={transitionDuration}>
      {(state) => {
        return (
          <div
            className={classNames(css.piece, css[pieceCode])}
            data-testid={`piece-${pieceCode}`}
            style={{
              ...getDefaultStyle(xYCoordinates),
              ...getTransitionStyles(transitionFrom, state, transitionDuration),
            }}
          ></div>
        );
      }}
    </Transition>
  );
};
