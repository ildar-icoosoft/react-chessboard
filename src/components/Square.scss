.square {
  cursor: pointer;
  display: flex;
  justify-content: center;
}

.dark {
  background-color: rgb(181, 136, 99);

  &.isDragging {
    background-color: #646f40;
  }
  &.isOver {
    background-color: #a5b56b;
  }
}

.light {
  background-color: rgb(240, 217, 181);

  &.isDragging {
    background-color: #829769;
  }
  &.isOver {
    background-color: #cbdf85;
  }
}

.isDragging :global(.piece) {
  opacity: 0.3;
}
