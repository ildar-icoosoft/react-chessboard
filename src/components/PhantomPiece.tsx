import React, { FC, useEffect, useState } from "react";
import { PieceCode } from "../enums/PieceCode";
import { Piece } from "./Piece";
import { DEFAULT_TRANSITION_DURATION } from "../constants/constants";
import css from "./PhantomPiece.scss";

export interface PhantomPieceProps {
  pieceCode: PieceCode;
  width?: number;
  transitionDuration?: number;
}

export const PhantomPiece: FC<PhantomPieceProps> = ({
  pieceCode,
  width,
  transitionDuration = DEFAULT_TRANSITION_DURATION,
}) => {
  const [showPhantom, setShowPhantom] = useState<boolean>(true);

  useEffect(() => {
    const timeoutId: NodeJS.Timeout = setTimeout(() => {
      setShowPhantom(false);
    }, transitionDuration);

    return () => clearTimeout(timeoutId);
  }, []);

  if (!showPhantom) {
    return null;
  }

  return (
    <div
      data-testid={`phantom-piece-${pieceCode}`}
      className={css.phantomPiece}
    >
      <Piece pieceCode={pieceCode} width={width} />
    </div>
  );
};
