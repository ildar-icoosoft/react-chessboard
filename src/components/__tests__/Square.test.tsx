import React, { createRef } from "react";
import TestRenderer from "react-test-renderer";
import { Square, SquareRef } from "../Square";
import { Piece } from "../Piece";
import { PieceCode } from "../../enums/PieceCode";
import { Notation } from "../Notation";
import { PieceColor } from "../../enums/PieceColor";
import { render, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { ReactDndRefType } from "../../interfaces/ReactDndRefType";
import { ITestBackend } from "react-dnd-test-backend";
import { DragDropManager, Identifier } from "dnd-core";
import { wrapInTestContext } from "react-dnd-test-utils";
import { PhantomPiece } from "../PhantomPiece";

jest.useFakeTimers();

describe("Square", () => {
  const SquareWithDnd = wrapInTestContext(Square);

  it("Snapshot", () => {
    const tree = TestRenderer.create(
      <SquareWithDnd coordinates={"a1"} pieceCode={PieceCode.BLACK_QUEEN} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    it("contains 1 Piece", () => {
      const testInstance = TestRenderer.create(
        <SquareWithDnd coordinates={"a1"} pieceCode={PieceCode.BLACK_QUEEN} />
      ).root;

      expect(() => testInstance.findByType(Piece)).not.toThrow();
    });

    it("does not contain a Piece", () => {
      const testInstance = TestRenderer.create(
        <SquareWithDnd coordinates={"a1"} />
      ).root;

      expect(() => testInstance.findByType(Piece)).toThrow();
    });

    it("contains 1 PhantomPiece", () => {
      const testInstance = TestRenderer.create(
        <SquareWithDnd
          coordinates={"a1"}
          pieceCode={PieceCode.BLACK_QUEEN}
          transitionFrom={{
            algebraic: "e2",
            x: 0,
            y: 0,
            phantomPiece: PieceCode.WHITE_BISHOP,
          }}
        />
      ).root;

      expect(() => testInstance.findByType(PhantomPiece)).not.toThrow();
    });

    it("contains 1 Notation", () => {
      const testRenderer = TestRenderer.create(
        <SquareWithDnd coordinates={"a1"} />
      );
      const testInstance = testRenderer.root;

      expect(() => testInstance.findByType(Notation)).not.toThrow();

      testRenderer.update(
        <SquareWithDnd coordinates={"a1"} showNotation={false} />
      );

      expect(() => testInstance.findByType(Notation)).toThrow();
    });
  });

  describe("children components props", () => {
    describe("Piece", () => {
      it("pieceCode", () => {
        const testInstance = TestRenderer.create(
          <SquareWithDnd coordinates={"a1"} pieceCode={PieceCode.BLACK_QUEEN} />
        ).root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.pieceCode).toBe(PieceCode.BLACK_QUEEN);
      });

      it("width", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd coordinates={"a1"} pieceCode={PieceCode.BLACK_QUEEN} />
        );
        const testInstance = testRenderer.root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.width).toBe(60);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            width={30}
          />
        );
        expect(piece.props.width).toBe(30);
      });

      it("transitionDuration", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd coordinates={"a1"} pieceCode={PieceCode.BLACK_QUEEN} />
        );
        const testInstance = testRenderer.root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );

        expect(piece.props.transitionDuration).toBeUndefined();

        testRenderer.update(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionDuration={600}
          />
        );

        expect(piece.props.transitionDuration).toBe(600);
      });

      it("transitionFrom", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd coordinates={"e4"} pieceCode={PieceCode.BLACK_QUEEN} />
        );
        const testInstance = testRenderer.root;

        let piece: TestRenderer.ReactTestInstance;

        piece = testInstance.findByType(Piece);
        expect(piece.props.transitionFrom).toBeUndefined();

        testRenderer.update(
          <SquareWithDnd
            coordinates={"e4"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
            }}
          />
        );

        piece = testInstance.findByType(Piece);
        expect(piece.props.transitionFrom).toEqual({
          algebraic: "e2",
          x: 0,
          y: 0,
        });
      });
    });

    describe("PhantomPiece", () => {
      it("pieceCode", () => {
        const testInstance = TestRenderer.create(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
          />
        ).root;

        const phantomPiece: TestRenderer.ReactTestInstance = testInstance.findByType(
          PhantomPiece
        );
        expect(phantomPiece.props.pieceCode).toBe(PieceCode.WHITE_ROOK);
      });

      it("width", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
          />
        );
        const testInstance = testRenderer.root;

        const phantomPiece: TestRenderer.ReactTestInstance = testInstance.findByType(
          PhantomPiece
        );
        expect(phantomPiece.props.width).toBe(60);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            width={30}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
          />
        );

        expect(phantomPiece.props.width).toBe(30);
      });

      it("transitionDuration", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
          />
        );
        const testInstance = testRenderer.root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          PhantomPiece
        );

        expect(piece.props.transitionDuration).toBeUndefined();

        testRenderer.update(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
            transitionDuration={600}
          />
        );

        expect(piece.props.transitionDuration).toBe(600);
      });
    });

    describe("Notation", () => {
      it("coordinates and orientation", () => {
        const testInstance = TestRenderer.create(
          <SquareWithDnd coordinates={"a1"} orientation={PieceColor.BLACK} />
        ).root;

        expect(() => {
          testInstance.find(
            (node) =>
              node.type === Notation &&
              node.props.coordinates === "a1" &&
              node.props.orientation === PieceColor.BLACK
          );
        }).not.toThrow();
      });

      it("width", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd coordinates={"a1"} orientation={PieceColor.BLACK} />
        );

        const testInstance = testRenderer.root;

        const notation: TestRenderer.ReactTestInstance = testInstance.findByType(
          Notation
        );
        expect(notation.props.width).toBe(60);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            width={30}
          />
        );
        expect(notation.props.width).toBe(30);
      });
    });
  });

  describe("children components remounting", () => {
    describe("Piece", () => {
      it("remounts if transitionFrom.algebraic is changed", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd coordinates={"e4"} pieceCode={PieceCode.BLACK_QUEEN} />
        );
        const testInstance = testRenderer.root;

        let piece: TestRenderer.ReactTestInstance;
        piece = testInstance.findByType(Piece);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"e4"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
            }}
          />
        );

        // old piece is unmounted because transitionFrom is changed (from undefined to object)
        expect(testInstance.findAll((item) => item === piece).length).toBe(0);

        piece = testInstance.findByType(Piece);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"e4"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "h8",
              x: 0,
              y: 0,
            }}
          />
        );

        // old piece is unmounted because transitionFrom.algebraic is changed (from e2 to h8)
        expect(testInstance.findAll((item) => item === piece).length).toBe(0);

        piece = testInstance.findByType(Piece);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"e8"}
            pieceCode={PieceCode.WHITE_PAWN}
            transitionFrom={{
              algebraic: "h8",
              x: 100,
              y: 200,
            }}
          />
        );

        // old piece is not unmounted because transitionFrom.algebraic is not changed
        expect(testInstance.findAll((item) => item === piece).length).toBe(1);
      });
    });

    describe("PhantomPiece", () => {
      it("remounts if transitionFrom.algebraic is changed", () => {
        const testRenderer = TestRenderer.create(
          <SquareWithDnd
            coordinates={"e4"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "e2",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
          />
        );
        const testInstance = testRenderer.root;

        let phantomPiece: TestRenderer.ReactTestInstance;
        phantomPiece = testInstance.findByType(PhantomPiece);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"e4"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "h1",
              x: 0,
              y: 0,
              phantomPiece: PieceCode.WHITE_ROOK,
            }}
          />
        );

        // old piece is unmounted because transitionFrom is changed (from e2 to h1)
        expect(
          testInstance.findAll((item) => item === phantomPiece).length
        ).toBe(0);

        phantomPiece = testInstance.findByType(PhantomPiece);

        testRenderer.update(
          <SquareWithDnd
            coordinates={"a1"}
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionFrom={{
              algebraic: "h1",
              x: 100,
              y: 200,
              phantomPiece: PieceCode.WHITE_KNIGHT,
            }}
          />
        );

        // old piece is not unmounted because transitionFrom.algebraic is not changed
        expect(
          testInstance.findAll((item) => item === phantomPiece).length
        ).toBe(1);
      });
    });
  });

  describe("methods", () => {
    it("getDragHandlerId()", () => {
      const dragAndDropRef = createRef<ReactDndRefType>();

      TestRenderer.create(
        <SquareWithDnd ref={dragAndDropRef} coordinates={"a1"} />
      );

      const squareRef: SquareRef = (dragAndDropRef.current as ReactDndRefType).getDecoratedComponent<
        SquareRef
      >();

      expect(() => squareRef.getDragHandlerId()).toBeTruthy();
    });

    it("getDropHandlerId()", () => {
      const dragAndDropRef = createRef<ReactDndRefType>();

      TestRenderer.create(
        <SquareWithDnd ref={dragAndDropRef} coordinates={"a1"} />
      );

      const squareRef: SquareRef = (dragAndDropRef.current as ReactDndRefType).getDecoratedComponent<
        SquareRef
      >();

      expect(() => squareRef.getDropHandlerId()).toBeTruthy();
    });

    it("getXYCoordinates()", () => {
      const dragAndDropRef = createRef<ReactDndRefType>();

      render(<SquareWithDnd ref={dragAndDropRef} coordinates={"a1"} />);

      const squareRef: SquareRef = (dragAndDropRef.current as ReactDndRefType).getDecoratedComponent<
        SquareRef
      >();

      expect(squareRef.getXYCoordinates()).toEqual({
        x: 0,
        y: 0,
      });
    });
  });

  describe("DOM structure", () => {
    it("Square color", () => {
      const { queryByTestId, rerender } = render(
        <SquareWithDnd coordinates={"a1"} />
      );
      expect(queryByTestId("square-a1")).toHaveClass("dark");

      rerender(<SquareWithDnd coordinates={"a2"} />);
      expect(queryByTestId("square-a2")).toHaveClass("light");

      rerender(<SquareWithDnd coordinates={"e5"} />);
      expect(queryByTestId("square-e5")).toHaveClass("dark");

      rerender(<SquareWithDnd coordinates={"g4"} />);
      expect(queryByTestId("square-g4")).toHaveClass("light");
    });

    it("must contain a div element in the root", () => {
      const { queryByTestId } = render(<SquareWithDnd coordinates={"a2"} />);
      expect(queryByTestId("square-a2")).toBeInTheDocument();
    });

    it("cssClass prop in div classList", () => {
      const { queryByTestId } = render(
        <SquareWithDnd coordinates={"a2"} cssClass={["class1", "class2"]} />
      );
      expect(queryByTestId("square-a2")).toHaveClass("class1", "class2");
    });
  });

  describe("Events", () => {
    it("Click", () => {
      const onSquareClick = jest.fn();
      const onSquareRightClick = jest.fn();

      const { getByTestId } = render(
        <SquareWithDnd
          coordinates={"a1"}
          onSquareClick={onSquareClick}
          onSquareRightClick={onSquareRightClick}
        />
      );

      fireEvent.click(getByTestId("square-a1"));

      expect(onSquareClick).toHaveBeenCalledTimes(1);
      expect(onSquareClick).toBeCalledWith("a1");

      expect(onSquareRightClick).toHaveBeenCalledTimes(0);
    });

    it("Click if no callback", () => {
      const { getByTestId } = render(<SquareWithDnd coordinates={"a1"} />);
      expect(() => {
        fireEvent.click(getByTestId("square-a1"));
      }).not.toThrow();
    });

    it("Right Click", () => {
      const onSquareClick = jest.fn();
      const onSquareRightClick = jest.fn();

      const { getByTestId } = render(
        <SquareWithDnd
          coordinates={"a1"}
          onSquareClick={onSquareClick}
          onSquareRightClick={onSquareRightClick}
        />
      );

      fireEvent.contextMenu(getByTestId("square-a1"));

      expect(onSquareRightClick).toHaveBeenCalledTimes(1);
      expect(onSquareRightClick).toBeCalledWith("a1");

      expect(onSquareClick).toHaveBeenCalledTimes(0);
    });

    it("Right Click if no callback", () => {
      const { getByTestId } = render(<SquareWithDnd coordinates={"a1"} />);
      expect(() => {
        fireEvent.contextMenu(getByTestId("square-a1"));
      }).not.toThrow();
    });

    it("onMouseEnterSquare", () => {
      const onMouseEnterSquare = jest.fn();

      const { getByTestId } = render(
        <SquareWithDnd
          coordinates={"a1"}
          onMouseEnterSquare={onMouseEnterSquare}
        />
      );

      fireEvent.mouseOver(getByTestId("square-a1"));

      expect(onMouseEnterSquare).toHaveBeenCalledTimes(1);
      expect(onMouseEnterSquare).toBeCalledWith("a1");
    });

    it("onMouseEnterSquare if no callback", () => {
      const { getByTestId } = render(<SquareWithDnd coordinates={"a1"} />);
      expect(() => {
        fireEvent.mouseOver(getByTestId("square-a1"));
      }).not.toThrow();
    });

    it("onMouseLeaveSquare", () => {
      const onMouseLeaveSquare = jest.fn();

      const { getByTestId } = render(
        <SquareWithDnd
          coordinates={"a1"}
          onMouseLeaveSquare={onMouseLeaveSquare}
        />
      );

      fireEvent.mouseOut(getByTestId("square-a1"));

      expect(onMouseLeaveSquare).toHaveBeenCalledTimes(1);
      expect(onMouseLeaveSquare).toBeCalledWith("a1");
    });

    it("onMouseLeaveSquare if no callback", () => {
      const { getByTestId } = render(<SquareWithDnd coordinates={"a1"} />);
      expect(() => {
        fireEvent.mouseOut(getByTestId("square-a1"));
      }).not.toThrow();
    });
  });

  describe("Drag and Drop", () => {
    describe("Begin Drag", () => {
      it("Cannot drag if draggable === false", () => {
        const ref = createRef<ReactDndRefType>();

        render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={false}
            pieceCode={PieceCode.WHITE_QUEEN}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;

        expect(manager.getMonitor().canDragSource(dragSourceId)).toBeFalsy();
      });

      it("Cannot drag if no piece", () => {
        const ref = createRef<ReactDndRefType>();

        render(<SquareWithDnd ref={ref} coordinates={"a2"} draggable={true} />);

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;

        expect(manager.getMonitor().canDragSource(dragSourceId)).toBeFalsy();
      });

      it("Can drag if contains piece and draggable === true", () => {
        const ref = createRef<ReactDndRefType>();

        render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={true}
            pieceCode={PieceCode.WHITE_QUEEN}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;

        expect(manager.getMonitor().canDragSource(dragSourceId)).toBeTruthy();
      });

      it("onDragStart event", () => {
        const onDragStart = jest.fn();

        const ref = createRef<ReactDndRefType>();

        render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={true}
            pieceCode={PieceCode.WHITE_QUEEN}
            onDragStart={onDragStart}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;

        const backend: ITestBackend = manager.getBackend() as ITestBackend;

        act(() => {
          backend.simulateBeginDrag([dragSourceId]);
        });

        expect(onDragStart).toHaveBeenCalledTimes(1);
        expect(onDragStart).toBeCalledWith({
          coordinates: "a2",
          pieceCode: PieceCode.WHITE_QUEEN,
        });

        act(() => {
          backend.simulateEndDrag();
        });
      });

      it("CSS classes", () => {
        const ref = createRef<ReactDndRefType>();

        const { queryByTestId } = render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={true}
            dragStartCssClass={"drag-start"}
            pieceCode={PieceCode.WHITE_QUEEN}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;

        expect(queryByTestId("square-a2")).not.toHaveClass("drag-start");

        const backend: ITestBackend = manager.getBackend() as ITestBackend;

        act(() => {
          backend.simulateBeginDrag([dragSourceId]);
        });

        expect(queryByTestId("square-a2")).toHaveClass("drag-start");

        act(() => {
          backend.simulateEndDrag();
        });
      });
    });

    describe("Drag over", () => {
      it("onDragEnterSquare event", () => {
        const onDragEnterSquare = jest.fn();

        const ref = createRef<ReactDndRefType>();

        render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={true}
            pieceCode={PieceCode.WHITE_QUEEN}
            onDragEnterSquare={onDragEnterSquare}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;
        const dropSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDropHandlerId() as Identifier;

        const backend: ITestBackend = manager.getBackend() as ITestBackend;

        act(() => {
          backend.simulateBeginDrag([dragSourceId]);
          backend.simulateHover([dropSourceId]);
        });

        expect(onDragEnterSquare).toHaveBeenCalledTimes(1);
        expect(onDragEnterSquare).toBeCalledWith("a2");

        act(() => {
          backend.simulateEndDrag();
        });
      });

      describe("CSS classes", () => {
        const ref = createRef<ReactDndRefType>();

        const { queryByTestId } = render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={true}
            dragEnterSquareCssClass={"drag-enter"}
            pieceCode={PieceCode.WHITE_QUEEN}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;
        const dropSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDropHandlerId() as Identifier;

        expect(queryByTestId("square-a2")).not.toHaveClass("drag-enter");

        const backend: ITestBackend = manager.getBackend() as ITestBackend;

        act(() => {
          backend.simulateBeginDrag([dragSourceId]);
          backend.simulateHover([dropSourceId]);
        });

        expect(queryByTestId("square-a2")).toHaveClass("drag-enter");

        act(() => {
          backend.simulateEndDrag();
        });
      });
    });

    describe("Drop", () => {
      it("onDrop event", () => {
        const onDrop = jest.fn();

        const ref = createRef<ReactDndRefType>();

        render(
          <SquareWithDnd
            ref={ref}
            coordinates={"a2"}
            draggable={true}
            pieceCode={PieceCode.WHITE_QUEEN}
            onDrop={onDrop}
          />
        );

        const manager: DragDropManager = (ref.current as ReactDndRefType).getManager() as DragDropManager;

        const dragSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDragHandlerId() as Identifier;
        const dropSourceId: Identifier = (ref.current as ReactDndRefType)
          .getDecoratedComponent<SquareRef>()
          .getDropHandlerId() as Identifier;

        const backend: ITestBackend = manager.getBackend() as ITestBackend;

        act(() => {
          backend.simulateBeginDrag([dragSourceId]);
          backend.simulateHover([dropSourceId]);
          backend.simulateDrop();
        });

        expect(onDrop).toHaveBeenCalledTimes(1);
        expect(onDrop).toBeCalledWith({
          sourceCoordinates: "a2",
          targetCoordinates: "a2",
          pieceCode: PieceCode.WHITE_QUEEN,
        });

        act(() => {
          backend.simulateEndDrag();
        });
      });
    });
  });
});
