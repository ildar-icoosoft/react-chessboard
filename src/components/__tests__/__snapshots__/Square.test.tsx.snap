// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square Snapshot 1`] = `
<div
  className="square dark"
  data-testid="square-a1"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "height": "60px",
      "width": "60px",
    }
  }
>
  <div
    className="piece bQ"
    data-testid="piece-bQ"
    style={Object {}}
  />
  <div
    className="rankNotation"
    data-testid="notation-rank-1"
    style={
      Object {
        "fontSize": "10px",
        "paddingRight": "50px",
      }
    }
  >
    1
  </div>
  <div
    className="fileNotation"
    data-testid="notation-file-a"
    style={
      Object {
        "fontSize": "10px",
        "paddingLeft": "50px",
      }
    }
  >
    a
  </div>
</div>
`;
